File: lib/seize/commands.rb
require 'json'

require 'seize/twitch'
require 'seize/utils'

module Seize 
  def self.watch(channel, quality)
    exec("livestreamer https://twitch.tv/#{channel} #{quality}")
  end

  def self.list(number)
    line_length = 79
    number_length = 5
    name_length = 20
    viewers_length = 12
    game_length = 40
    row_template = " %-#{number_length}{number}" \
                   " %-#{name_length}{name}" \
                   " %-#{viewers_length}{viewers}" \
                   " %{game}\n"
    result = ''
    line = '-' * line_length << "\n"
    result << line
    result << row_template % {
      number: '#',
      name: 'channel',
      viewers: 'viewers',
      game: 'game'
    } 
    result << line
    streams = Seize::Twitch.get_streams(limit: number)['streams']
    streams.each_with_index do |stream, i|
      result << row_template % {
        number: "#{i+1}.",
        name: Seize::Utils.truncate(stream['channel']['name'], name_length),
        viewers: Seize::Utils.commaize(stream['viewers']),
        game: Seize::Utils.truncate(stream['game'], game_length)
      }
    end
    puts result
  end

  def self.check(channel_name)
    channel = Seize::Twitch.get_channel(channel_name)
    if channel.nil?
      puts 'error: channel does not exist'
    elsif channel['stream'].nil?
      puts "#{channel_name} is offline"
    else
      game = channel['stream']['game']
      viewers = Seize::Utils.commaize(channel['stream']['viewers'])
      puts "#{channel_name} is playing #{game} with #{viewers} viewers"
    end
  end
end

File: src/commands.js
import _ from 'lodash';
import childProcess from 'child_process';

import twitch from './twitch';

async function list(number = null, game = null) {
  const seperator = '-'.repeat(80);
  let result = [];
  result.push(seperator);
  result.push(rowTemplate('#', 'name', 'viewers', 'game'));
  result.push(seperator);
  const streams = await twitch.getStreams(number, game);
  const rows = streams.map((stream, i) => {
    const { channel: { name }, viewers, game } = stream;
    return rowTemplate(i + 1, name, viewers, game);
  });
  return result.concat(rows).join('\n');
}

function rowTemplate(number, name, viewers, game) {
  const numberColumns = 6;
  const nameColumns = 21;
  const viewersColumns = 13;
  const gameColumns = 40;
  const paddedNumber = _.padEnd(
    _.truncate(number, { length: numberColumns }), numberColumns);
  const paddedName = _.padEnd(
    _.truncate(name, { length: nameColumns }), nameColumns);
  const paddedViewers = _.padEnd(
    typeof(number) === 'number'
      ? _.truncate(
        new Intl.NumberFormat().format(viewers), {
          length: viewersColumns
        })
      : viewers, viewersColumns);
  const truncatedGame = _.truncate(game, { length: gameColumns });
  return ` ${paddedNumber}${paddedName}${paddedViewers}${truncatedGame}`;
}

async function check(channelName) {
  const streams = await twitch.searchStreams(channelName);
  const found = streams.find(stream => {
    return stream.channel.display_name === channelName;
  });
  if (found === undefined) {
    return `${channelName} is offline or doesn't exist`;
  }
  const game = found['game'];
  const viewers = new Intl.NumberFormat().format(found['viewers']);
  return `${channelName} is playing ${game} with ${viewers} viewers`;
}

function watch(channel, quality) {
  return childProcess.spawn(
    'livestreamer', [`https://twitch.tv/${channel}`, quality]);
}

export default { list, watch, check };
